cql3/single_column_relation.cc

```diff
@@ -67,32 +68,47 @@ single_column_relation::to_term(const std::vector<lw_shared_ptr<column_specifica
+    using namespace wip;
     const column_definition& column_def = to_column_definition(*schema, *_entity);
     if (!_map_key) {
         auto term = to_term(to_receivers(*schema, column_def), *_value, db, schema->ks_name(), bound_names);
-        return ::make_shared<single_column_restriction::EQ>(column_def, std::move(term));
+        auto restr = ::make_shared<single_column_restriction::EQ>(column_def, term);
+        restr->expression = binary_operator{std::vector{column_value(&column_def)}, &operator_type::EQ, term};
+        return restr;
     }
     auto&& receivers = to_receivers(*schema, column_def);
     auto&& entry_key = to_term({receivers[0]}, *_map_key, db, schema->ks_name(), bound_names);
     auto&& entry_value = to_term({receivers[1]}, *_value, db, schema->ks_name(), bound_names);
-    return make_shared<single_column_restriction::contains>(column_def, std::move(entry_key), std::move(entry_value));
+    auto restr = make_shared<single_column_restriction::contains>(column_def, entry_key, entry_value);
+    restr->expression =
+            binary_operator{std::vector{column_value(&column_def, entry_key)}, &operator_type::EQ, entry_value};
+    return restr;
 }
 
 ::shared_ptr<restrictions::restriction>
 single_column_relation::new_IN_restriction(database& db, schema_ptr schema, variable_specifications& bound_names) {
+    using namespace restrictions::wip;
     const column_definition& column_def = to_column_definition(*schema, *_entity);
     auto receivers = to_receivers(*schema, column_def);
     assert(_in_values.empty() || !_value);
     if (_value) {
         auto term = to_term(receivers, *_value, db, schema->ks_name(), bound_names);
-        return make_shared<single_column_restriction::IN_with_marker>(column_def, dynamic_pointer_cast<lists::marker>(term));
+        auto r = make_shared<single_column_restriction::IN_with_marker>(
+                column_def, dynamic_pointer_cast<lists::marker>(term));
+        r->expression = binary_operator{std::vector{column_value(&column_def)}, &operator_type::IN, std::move(term)};
+        return r;
     }
     auto terms = to_terms(receivers, _in_values, db, schema->ks_name(), bound_names);
     // Convert a single-item IN restriction to an EQ restriction
     if (terms.size() == 1) {
-        return ::make_shared<single_column_restriction::EQ>(column_def, std::move(terms[0]));
+        auto restr = ::make_shared<single_column_restriction::EQ>(column_def, terms[0]);
+        restr->expression = binary_operator{std::vector{column_value(&column_def)}, &operator_type::EQ, terms[0]};
+        return restr;
     }
-    return ::make_shared<single_column_restriction::IN_with_values>(column_def, std::move(terms));
+    auto r = ::make_shared<single_column_restriction::IN_with_values>(column_def, terms);
+    r->expression = binary_operator{std::vector{column_value(&column_def)}, &operator_type::IN,
+        ::make_shared<lists::delayed_value>(std::move(terms))};
+    return r;
```

Comment:
Switching to passing expression via constructor will simplify this and similar code.

Reply:
Agreed, though this will all necessarily disappear when I drop the old code.  Instead of `r->expression = xyz` it will just say `return xyz`, because the return type will be `expression`.